### PR TITLE
Create foster account3

### PR DESCRIPTION
**Pull request by ErinClaudio**
# 🔗 Issue
https://github.com/rubyforgood/pet-rescue/issues/449

# ✍️ Description

# 📷 Screenshots/Demos

### CODE REVIEW

**Comment by ErinClaudio** [comment]
@kasugaijin I have more to do but I think this is closer to what you wanted.

File: db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class RemoveProfilesAndAddParentFromAdopterFosterProfiles < ActiveRecord::Migration[7.0]
+  def change
+    remove_column :adopter_foster_profiles, :adopter_account_id, :bigint
+    remove_column :adopter_foster_profiles, :foster_account_id, :bigint
```
Comment (kasugaijin) [inline db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb:4]: is this removal needed in this particular branch? I had it because I was on your branch where you had already added this column in a migration.

**Comment by kasugaijin** [comment]
yes this looks like it! I think all is left is a switch and replace for any cases of adopter_foster_profile.adopter_account now is adopter_foster_profile.parent

**Comment by ErinClaudio** [comment]
@kasugaijin Some changes made. tests failing. I will be in the meeting tonight

File: db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb (hunk 1 above)
Comment (ErinClaudio) [reply db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb:4]: I added this back in

**Comment by kasugaijin** [comment]
@ErinClaudio Hmm weird - [my PR has 8 errors](https://github.com/rubyforgood/pet-rescue/actions/runs/8196586683/job/22417120134?pr=502) vs 28 errors in this one so there must be a difference in code that's causing that?

File: db/migrate/20240312150459_add_back_columns_to_adopter_foster_profiles.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class AddBackColumnsToAdopterFosterProfiles < ActiveRecord::Migration[7.0]
```
Comment (kasugaijin) [inline db/migrate/20240312150459_add_back_columns_to_adopter_foster_profiles.rb:1]: Do we need this migration? adopter_account_id is already on adopter_foster_profiles table. Then we have a migration above to remove both. 

I think it makes more sense to just remove the adopter_account_id from adopter_foster_profiles.

File: db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb (hunk 1 above)
Comment (kasugaijin) [reply db/migrate/20240311173819_remove_profiles_and_add_parent_from_adopter_foster_profiles.rb:4]: Ah. Why is that?

File: db/migrate/20240312150459_add_back_columns_to_adopter_foster_profiles.rb (hunk 2 above)
Comment (ErinClaudio) [reply db/migrate/20240312150459_add_back_columns_to_adopter_foster_profiles.rb:1]: agreed

**Comment by nsiwnf** [comment]
@ErinClaudio let's use the term `caretaker` instead of `parent` since "parent" has a different meaning in programming

**Comment by ErinClaudio** [comment]
I am going to reset this branch and run the migrations without the residual code. I feel like this will leave the code in a more maintainable/cleaner state.

File: app/models/adopter_foster_account.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#
+#  index_adopter_foster_accounts_on_user_id  (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (user_id => users.id)
+#
+class AdopterFosterAccount < ApplicationRecord
+  belongs_to :user
+  has_one :adopter_foster_profile, as: :caretaker
```
Comment (ErinClaudio) [inline app/models/adopter_foster_account.rb:20]: I added this in per this [work](https://github.com/rubyforgood/pet-rescue/blob/2bc73b48781e6b432d059a7d8dc1d7f412b24ea0/db/migrate/20240304183316_add_foster_account_id_to_adopter_foster_profiles.rb) @ben. not sure if it is needed.

**Comment by ErinClaudio** [comment]
Ok, I am happy to continue working on this. But before continuing on I wanted to make sure that this current stopping point is correct. Thank you for the feedback @kasugaijin @mononoken @nsiwnf 

next steps resolving the seeding issue and tests( which we did earlier)

**Comment by ErinClaudio** [comment]
My [notes](https://gist.github.com/ErinClaudio/822992ce207da5e014ae1d08f8b079fe)

File: app/models/adopter_foster_account.rb (hunk 3 above)
Comment (kasugaijin) [reply app/models/adopter_foster_account.rb:20]: So we can remove the as caretaker now as we no longer need the polymorphic association. It's a direct association now where

`AdopterFosterAccount` `has_one :adopter_foster_profile`
`AdopterFosterProfile` `belongs_to :adopter_foster_account`

File: app/models/adopter_foster_profile.rb [hunk 4]
```diff
@@ -2,57 +2,63 @@
...
 # Foreign Keys
 #
-#  fk_rails_...  (adopter_account_id => adopter_accounts.id)
+#  fk_rails_...  (adopter_foster_account_id => adopter_foster_accounts.id)
 #  fk_rails_...  (location_id => locations.id)
 #
 class AdopterFosterProfile < ApplicationRecord
   belongs_to :location, dependent: :destroy
   belongs_to :adopter_account
+  belongs_to :caretaker, polymorphic: true, optional: true
```
Comment (kasugaijin) [inline app/models/adopter_foster_profile.rb:61]: So this now is `belongs_to :adopter_foster_account`

File: db/migrate/20240314171529_create_adopter_foster_accounts.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class CreateAdopterFosterAccounts < ActiveRecord::Migration[7.0]
```
Comment (kasugaijin) [inline db/migrate/20240314171529_create_adopter_foster_accounts.rb:1]: Instead of creating a new table, we could just rename the `adopter_accounts` table to `adopter_foster_accounts`

File: app/models/user.rb [hunk 6]
```diff
@@ -58,6 +58,7 @@ class User < ApplicationRecord
 
   has_one :staff_account, dependent: :destroy
   has_one :adopter_account, dependent: :destroy
+  has_one :adopter_foster_profile, dependent: :destroy
```
Comment (kasugaijin) [inline app/models/user.rb:61]: rename line 60 to `adopter_foster_account` and delete line 61

File: db/schema.rb [hunk 7]
```diff
@@ -98,8 +97,14 @@
     t.text "visit_dates"
     t.text "referral_source"
     t.bigint "location_id", null: false
-    t.index ["adopter_account_id"], name: "index_adopter_foster_profiles_on_adopter_account_id"
+    t.string "parent_type"
+    t.bigint "parent_id"
+    t.bigint "adopter_account_id"
+    t.bigint "foster_account_id"
+    t.bigint "adopter_foster_account_id"
```
Comment (kasugaijin) [inline db/schema.rb:104]: 101, 102 and 103 are no longer required on this table

**Comment by kasugaijin** [comment]
> Ok, I am happy to continue working on this. But before continuing on I wanted to make sure that this current stopping point is correct. Thank you for the feedback @kasugaijin @mononoken @nsiwnf
> 
> next steps resolving the seeding issue and tests( which we did earlier)

It might be the easiest to just start from Main and make the changes. You don't have to, but it might be easier because you have a number of no longer required changes in that PR that you have to undo. Basically, what we need comparing changes to main, is to rename AdopterAccount to AdopterFosterAccount for the model, table, and code references. So a migration to rename the Adopter Accounts table to Adopter Foster Accounts table, and a migration to rename the adopter accounts id column on AdopterFosterProfiles table to adopter foster accounts id. Then a search and replace for any calls to .adopter_account in the code to .adopter_foster_account

**Comment by nsiwnf** [comment]
@ErinClaudio thanks for being flexible with this work - I know we're pivoting a bit and it's not ideal to have many redos. Agree with @kasugaijin's suggestion about starting from main and doing renames

**Comment by ErinClaudio** [comment]
Agreed I will start from a new branch. I will delete this PR and Branch after the new branch is merged.
